05-structs.Rmd

```diff
@@ -4,11 +4,47 @@ ottrpal::set_knitr_image_path()
 
 # Organizing variables via Structs
 
+In our workflow so far, we see that certain variables are always used together, even for different tasks. For example, variables related to the reference genome are always used for the same purpose and passed on to tasks in almost the same way. This leads to quite a bit of coding redundancy, as when we write down the large set of variables related to the reference genome as task inputs, we are just thinking about one entity. We don't make distinctions of the reference genome files until the task body itself.
 
+To improve code organization and readability, we can package all variables related to the reference genome into a compound data structure called a **struct**. With a struct variable, we can refer all the packaged variables as one single variable, and also refer to specific variables within the struct without losing any information.
 
-Our final code:
+To define a struct, we must declare it outside of a `workflow` and `task`:
 
+```         
+struct referenceGenome {
+    File ref_fasta
+    File ref_fasta_index
+    File ref_dict
+    File ref_amb
+    File ref_ann
+    File ref_bwt
+    File ref_pac
+    File ref_sa
+    String ref_name
+}
+
+workflow minidata_mutation_calling_v1 {
+  input {
+    File sampleFastq
+    referenceGenome refGenome
+    ...
+  }
+  # Map reads to reference
+  call BwaMem {
+    input:
+      input_fastq = sampleFastq,
+      refGenome = refGenome
+  }
+}
 ```
+
+The `referenceGenome` struct contains all the variables related to the reference genome, but values cannot be defined. The struct definition merely lays the skeleton components of the data structure, but contains no actual values.
```

Comment:
Consider changing The `referenceGenome` struct contains all the variables related to the reference genome, but values cannot be defined. 
to 
The `referenceGenome` struct contains all the variables related to the reference genome, but values cannot be defined here.